### DISCUSSION TITLE
Microsoft Security Advisory CVE-2023-33170: .NET Security Feature Bypass Vulnerability

### DISCUSSION THREAD

**Discussion by rbhanda** [dotnet/aspnetcore#64462] [Security advisories]
# Microsoft Security Advisory CVE-2023-33170: .NET Security Feature Bypass Vulnerability

## <a name="executive-summary"></a>Executive summary

Microsoft is releasing this security advisory to provide information about a vulnerability in ASP.NET Core 2.1 and above. This advisory also provides guidance on what developers can do to update their applications to remove this vulnerability.

A vulnerability exist in ASP.NET  Core applications where account lockout maximum failed attempts may not be immediately updated, allowing an attacker to try more passwords.

## Announcement

Announcement for this issue can be found at https://github.com/dotnet/announcements/issues/264

### <a name="mitigation-factors"></a>Mitigation factors

Microsoft has not identified any mitigating factors for this vulnerability.

## <a name="affected-software"></a>Affected software

* Any ASP.NET 7.0 application running on .NET 7.0.8 or earlier.
* Any ASP.NET 6.0 application running on .NET 6.0.19 or earlier.
* Any ASP.NET Core 2.1 application consuming the package Microsoft.AspNetCore.Identity version 2.1.31 or earlier.

If your application uses the following package versions, ensure you update to the latest version of .NET.

### <a name="ASP.NET Core 2.1"></a>ASP.NET Core 2.1

Package name | Affected version | Patched version
------------ | ---------------- | -------------------------
[ Microsoft.AspNetCore.Identity](https://www.nuget.org/packages/Microsoft.AspNetCore.Identity/2.1.39)      | <=2.1.39 | 2.1.39

### <a name="Microsoft.AspNet.Identity.Owin"></a>Microsoft.AspNet.Identity.Owin
Package name | Affected version | Patched version
------------ | ---------------- | -------------------------
[ Microsoft.AspNet.Identity.Owin](https://www.nuget.org/packages/Microsoft.AspNet.Identity.Owin)    |  <= 2.2.3 | 2.2.4

### <a name="ASP.NET 6"></a>ASP.NET 6.0

Package name | Affected version | Patched version
------------ | ---------------- | -------------------------

[Microsoft.AspNetCore.App.Runtime.linux-arm](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.linux-arm)               |  <= 6.0.19 | 6.0.20
[Microsoft.AspNetCore.App.Runtime.linux-arm64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.linux-arm64)           |  <= 6.0.19 | 6.0.20
[Microsoft.AspNetCore.App.Runtime.linux-musl-arm](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.linux-musl-arm)     |  <= 6.0.19 | 6.0.20
[Microsoft.AspNetCore.App.Runtime.linux-musl-arm64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.linux-musl-arm64) |  <= 6.0.19 | 6.0.20
[Microsoft.AspNetCore.App.Runtime.linux-musl-x64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.linux-musl-x64)     |  <= 6.0.19 | 6.0.20
[Microsoft.AspNetCore.App.Runtime.linux-x64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.linux-x64)               |  <= 6.0.19 | 6.0.20
[Microsoft.AspNetCore.App.Runtime.osx-arm64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.osx-arm64)               |  <= 6.0.19 | 6.0.20
[Microsoft.AspNetCore.App.Runtime.osx-x64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.osx-x64)                   |  <= 6.0.19 | 6.0.20
[Microsoft.AspNetCore.App.Runtime.win-arm](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.win-arm)                   |  <= 6.0.19 | 6.0.20
[Microsoft.AspNetCore.App.Runtime.win-arm64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.win-arm64)               |  <= 6.0.19 | 6.0.20
[Microsoft.AspNetCore.App.Runtime.win-x64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.win-x64)                   |  <= 6.0.19 | 6.0.20
[Microsoft.AspNetCore.App.Runtime.win-x86](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.win-x86)                   |  <= 6.0.19 | 6.0.20

### <a name="ASP.NET 7"></a>ASP.NET 7.0

Package name | Affected version | Patched version
------------ | ---------------- | -------------------------
[Microsoft.AspNetCore.App.Runtime.linux-arm](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.linux-arm)               |  <= 7.0.9 | 7.0.9
[Microsoft.AspNetCore.App.Runtime.linux-arm64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.linux-arm64)           |  <= 7.0.9 | 7.0.9
[Microsoft.AspNetCore.App.Runtime.linux-musl-arm](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.linux-musl-arm)     |  <= 7.0.9 | 7.0.9
[Microsoft.AspNetCore.App.Runtime.linux-musl-arm64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.linux-musl-arm64) |  <= 7.0.9 | 7.0.9
[Microsoft.AspNetCore.App.Runtime.linux-musl-x64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.linux-musl-x64)     |  <= 7.0.9 | 7.0.9
[Microsoft.AspNetCore.App.Runtime.linux-x64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.linux-x64)               |  <= 7.0.9 | 7.0.9
[Microsoft.AspNetCore.App.Runtime.osx-arm64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.osx-arm64)               |  <= 7.0.9 | 7.0.9
[Microsoft.AspNetCore.App.Runtime.osx-x64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.osx-x64)                   |  <= 7.0.9 | 7.0.9
[Microsoft.AspNetCore.App.Runtime.win-arm](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.win-arm)                   |  <= 7.0.9 | 7.0.9
[Microsoft.AspNetCore.App.Runtime.win-arm64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.win-arm64)               |  <= 7.0.9 | 7.0.9
[Microsoft.AspNetCore.App.Runtime.win-x64](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.win-x64)                   |  <= 7.0.9 | 7.0.9
[Microsoft.AspNetCore.App.Runtime.win-x86](https://www.nuget.org/packages/Microsoft.AspNetCore.App.Runtime.win-x86)                   |  <= 7.0.9 | 7.0.9

### <a name="how-affected"></a>How do I know if I am affected?

If you have a runtime or SDK with a version listed, or an affected package listed in [affected software](#affected-software), you're exposed to the vulnerability.

### <a name="how-fix"></a>How do I fix the issue?

* To fix the issue please install the latest version of .NET 6.0 or .NET 7.0. If you have installed one or more .NET SDKs through Visual Studio, Visual Studio will prompt you to update Visual Studio, which will also update your .NET  SDKs.
* If you are using one of the affected packages, please update to the patched version listed above.
* If you have .NET 6.0 or greater installed, you can list the versions you have installed by running the `dotnet --info` command. You will see output like the following;

```
.NET Core SDK (reflecting any global.json):

 Version:   6.0.300
 Commit:    8473146e7d

Runtime Environment:

 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.300\

Host (useful for support):

  Version: 6.0.5
  Commit:  8473146e7d

.NET Core SDKs installed:

  6.0.300 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:

  Microsoft.AspNetCore.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

* If you're using .NET 7.0, you should download and install Runtime 7.0.9 or SDK 7.0.109 (for Visual Studio 2022 v17.4) from https://dotnet.microsoft.com/download/dotnet-core/7.0.
* If you're using .NET 6.0, you should download and install Runtime 6.0.20 or SDK 6.0.315 (for Visual Studio 2022 v17.2) from https://dotnet.microsoft.com/download/dotnet-core/6.0.
* If you're using ASP.NET Core 2.1 and consuming the Microsoft.AspNetCore.Identity package, update that package to version [2.1.39](https://www.nuget.org/packages/Microsoft.AspNetCore.Identity/2.1.39)

.NET 6.0 and and .NET 7.0 updates are also available from Microsoft Update. To access this either type "Check for updates" in your Windows search, or open Settings, choose Update & Security and then click Check for Updates.

Once you have installed the updated runtime or SDK, restart your apps for the update to take effect.

Additionally, if you've deployed [self-contained applications](https://docs.microsoft.com/dotnet/core/deploying/#self-contained-deployments-scd) targeting any of the impacted versions, these applications are also vulnerable and must be recompiled and redeployed.

### Reporting Security Issues

If you have found a potential security issue in .NET 6.0 or .NET 7.0, please email details to secure@microsoft.com. Reports may qualify for the Microsoft .NET Core & .NET 5 Bounty. Details of the Microsoft .NET Bounty Program including terms and conditions are at <https://aka.ms/corebounty>.

### Support

You can ask questions about this issue on GitHub in the .NET GitHub organization. The main repos are located at https://github.com/dotnet/runtime and https://github.com/dotnet/aspnet/. The Announcements repo (https://github.com/dotnet/Announcements) will contain this bulletin as an issue and will include a link to a discussion issue. You can ask questions in the linked discussion issue.

### Disclaimer

The information provided in this advisory is provided "as is" without warranty of any kind. Microsoft disclaims all warranties, either express or implied, including the warranties of merchantability and fitness for a particular purpose. In no event shall Microsoft Corporation or its suppliers be liable for any damages whatsoever including direct, indirect, incidental, consequential, loss of business profits or special damages, even if Microsoft Corporation or its suppliers have been advised of the possibility of such damages. Some states do not allow the exclusion or limitation of liability for consequential or incidental damages so the foregoing limitation may not apply.

### External Links

[CVE-2023-33170]( https://www.cve.org/CVERecord?id=CVE-2023-33170)

### Revisions

V1.0 (July 11, 2023): Advisory published.

_Version 1.0_

_Last Updated 2023-07-11_

**Comment by johnkors** [comment]
>Affected software

>Any ASP.NET 7.0 application running on .NET 7.0.8 or earlier.

>If you're using .NET 7.0, you should download and install Runtime 7.0.7

🤔

**Comment by danmoseley** [comment]
> 🤔

I believe I corrected the post -- @rbhanda please confirm.

**Comment by Kahbazi** [comment]
@rbhanda The link for the announcement is the wrong issue. I believe this is the correct one. https://github.com/dotnet/announcements/issues/264

**Comment by rbhanda** [comment]
> @rbhanda The link for the announcement is the wrong issue. I believe this is the correct one. [dotnet/announcements#264](https://github.com/dotnet/announcements/issues/264)

Sorry about that. I corrected the links and the text in the [announcement ](https://github.com/dotnet/announcements/issues/264) and this issue

**Comment by phenning** [comment]
@rbhanda Can you confirm whether this vulnerability exists in [Microsoft.AspNet.Identity.Owin](https://www.nuget.org/packages/Microsoft.AspNet.Identity.Owin) when targeting .NET Framework TFMS (non NET Core)?

It's not super clear in the bullet points under affected software, even though it is listed in the table below.

**Comment by rbhanda** [comment]
> @rbhanda Can you confirm whether this vulnerability exists in [Microsoft.AspNet.Identity.Owin](https://www.nuget.org/packages/Microsoft.AspNet.Identity.Owin) when targeting .NET Framework TFMS (non NET Core)?
> 
> It's not super clear in the bullet points under affected software, even though it is listed in the table below.

@GrabYourPitchforks @halter73

**Comment by halter73** [comment]
Yes. It affects any consumer of Microsoft.AspNet.Identity.Owin using the account lockout feature including but not limited to full framework.

**Comment by phenning** [comment]
> Yes. It affects any consumer of Microsoft.AspNet.Identity.Owin using the account lockout feature including but not limited to full framework.

Thanks Stephen, wanted to make sure before we started updating the .NET Framework templates in Visual Studio to use the new version.

**Comment by 11qqaazzqwerty** [comment]
Is it really fixed in version 6.0.20?
Microsoft Defender for Cloud treat this version as affected:
![image](https://github.com/dotnet/aspnetcore/assets/121670476/33597acd-5f9e-4e18-8e58-7d285e90947f)

**Comment by peterlukerow** [comment]
@halter73 
> Yes. It affects any consumer of Microsoft.AspNet.Identity.Owin using the account lockout feature including but not limited to full framework.

I'm still not clear on this, it's all a little confusing. I'm using .NET 4.8 which is not listed as one of the affect .NETs. Here you state it doesn't matter which framework you are on it's the version of Microsoft.AspNet.Identity.Owin.

However, on [(https://github.com/github/advisory-database/pull/2496)] it says that for Microsoft.AspNet.Identity.Owin you need at least version 7.0.9, however in VS 2019 when managing nugets for the solution we are using v2.2.3 and it says that v2.2.4 is the latest stable, there is no v7.0.9 to upgrade to, thus my confusion.

In a recent external pen test of our web app and on anecdotal evidence an account is always locked out after the number of times configured in our SQL DB, i.e. after 5 invalid attempts, 6th attempt fails even if you use valid credentials.

**Comment by halter73** [comment]
v2.2.4 is the latest stable. https://github.com/https://github.com/github/advisory-database/pull/2496 is an outdated PR between feature branches. If you take a look at [what's currently in the `main` branch](https://github.com/github/advisory-database/blob/df30e8a3304bb2887c17f2b63a5c57131b527966/advisories/github-reviewed/2023/07/GHSA-25c8-p796-jg6r/GHSA-25c8-p796-jg6r.json#L37-L55), that indicates 2.2.4 is the patched version just like the announcement at https://github.com/dotnet/announcements/issues/264.
